File: client/go/traits.go
// Package angzarr provides handler interfaces (traits) for the unified router pattern.
//
// These interfaces define the contract for domain handlers. Implementors
// encapsulate their routing logic internally and declare which types
// they handle via CommandTypes(), EventTypes(), etc.
//
// This mirrors the Rust trait pattern from client/rust/src/router/traits.rs.
package angzarr

import (
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"google.golang.org/protobuf/types/known/anypb"
)

// ============================================================================
// Common Types
// ============================================================================

// RejectionHandlerResponse is defined in compensation.go.
// It contains optional Events (compensation) and Notification (upstream propagation).

// SagaHandlerResponse is the response from saga handlers.
type SagaHandlerResponse struct {
	// Commands to send to other aggregates.
	Commands []*pb.CommandBook
	// Facts/events to inject to other aggregates.
	Events []*pb.EventBook
}

// ProcessManagerResponse is the response from process manager handlers.
type ProcessManagerResponse struct {
	// Commands to send to other aggregates.
	Commands []*pb.CommandBook
	// Events to persist to the PM's own domain.
	ProcessEvents *pb.EventBook
	// Facts to inject to other aggregates.
	Facts []*pb.EventBook
}

// ============================================================================
// CommandHandler Handler Interface
// ============================================================================

// CommandHandlerDomainHandler is the handler for a single domain's commands.
//
// Command handlers receive commands and emit events. They maintain state
// that is rebuilt from events using a StateRouter.
//
// Example:
//
//	type PlayerHandler struct {
//	    stateRouter *angzarr.StateRouter[PlayerState]
//	}
//
//	func (h *PlayerHandler) CommandTypes() []string {
//	    return []string{"examples.RegisterPlayer", "examples.DepositFunds"}
//	}
//
//	func (h *PlayerHandler) StateRouter() *angzarr.StateRouter[PlayerState] {
//	    return h.stateRouter
//	}
//
//	func (h *PlayerHandler) Handle(
//	    cmd *pb.CommandBook,
//	    payload *anypb.Any,
//	    state *PlayerState,
//	    seq uint32,
//	) (*pb.EventBook, error) {
//	    // Dispatch to specific command handlers
//	}
type CommandHandlerDomainHandler[S any] interface {
	// CommandTypes returns the fully-qualified command type names this handler processes.
	// Used for subscription derivation and routing.
	CommandTypes() []string

	// Rebuild reconstructs state from events.
	Rebuild(events *pb.EventBook) S

	// Handle processes a command and returns resulting events.
	// The handler should dispatch internally based on payload.TypeUrl.
	Handle(
		cmd *pb.CommandBook,
		payload *anypb.Any,
		state S,
		seq uint32,
	) (*pb.EventBook, error)

	// OnRejected handles a rejection notification.
	// Called when a command issued by a saga/PM targeting this command handler's
	// domain was rejected. Override to provide custom compensation logic.
	// Default implementation should return an empty response.
	OnRejected(
		notification *pb.Notification,
		state S,
		targetDomain string,
		targetCommand string,
	) (*RejectionHandlerResponse, error)
}

// ============================================================================
// Saga Handler Interface
// ============================================================================

// SagaDomainHandler is the handler for a single domain's events in a saga.
//
// Sagas translate events from one domain into commands for another.
// They are stateless -- each event is processed independently.
//
// Example:
//
//	type OrderSagaHandler struct{}
//
//	func (h *OrderSagaHandler) EventTypes() []string {
//	    return []string{"examples.OrderCompleted", "examples.OrderCancelled"}
//	}
//
//	func (h *OrderSagaHandler) Prepare(source *pb.EventBook, event *anypb.Any) []*pb.Cover {
//	    // Declare needed destinations
//	}
//
//	func (h *OrderSagaHandler) Execute(
//	    source *pb.EventBook,
//	    event *anypb.Any,
//	    destinations []*pb.EventBook,
//	) (*SagaHandlerResponse, error) {
//	    // Produce commands and/or events
//	}
type SagaDomainHandler interface {
	// EventTypes returns the fully-qualified event type names this handler processes.
	// Used for subscription derivation.
	EventTypes() []string

	// Prepare declares destination covers needed.
	// Called before Execute to fetch destination aggregate state.
	Prepare(source *pb.EventBook, event *anypb.Any) []*pb.Cover

	// Execute produces commands and/or events.
	// Called with source event and fetched destination state.
	Execute(
		source *pb.EventBook,
		event *anypb.Any,
		destinations []*pb.EventBook,
	) (*SagaHandlerResponse, error)

	// OnRejected handles a rejection notification.
	// Called when a saga-issued command was rejected.
	// Default implementation should return an empty response.
	OnRejected(
		notification *pb.Notification,
		targetDomain string,
		targetCommand string,
	) (*RejectionHandlerResponse, error)
}

// ============================================================================
// Process Manager Handler Interface
// ============================================================================

// ProcessManagerDomainHandler is the handler for a single domain's events in a PM.
//
// Process managers correlate events across multiple domains and maintain
// their own state. Each domain gets its own handler, but they all share
// the same PM state type.
//
// Example:
//
//	type OrderPmHandler struct{}
//
//	func (h *OrderPmHandler) EventTypes() []string {
//	    return []string{"examples.OrderCreated"}
//	}
//
//	func (h *OrderPmHandler) Prepare(
//	    trigger *pb.EventBook,
//	    state *HandFlowState,
//	    event *anypb.Any,
//	) []*pb.Cover {
//	    // Declare needed destinations
//	}
//
//	func (h *OrderPmHandler) Handle(
//	    trigger *pb.EventBook,
//	    state *HandFlowState,
//	    event *anypb.Any,
//	    destinations []*pb.EventBook,
//	) (*ProcessManagerResponse, error) {
//	    // Process event, emit commands and/or PM events
//	}
type ProcessManagerDomainHandler[S any] interface {
	// EventTypes returns the fully-qualified event type names this handler processes.
	EventTypes() []string

	// Prepare declares destination covers needed.
	Prepare(trigger *pb.EventBook, state S, event *anypb.Any) []*pb.Cover

	// Handle processes the event and produces commands and PM events.
	Handle(
		trigger *pb.EventBook,
		state S,
		event *anypb.Any,
		destinations []*pb.EventBook,
	) (*ProcessManagerResponse, error)

	// OnRejected handles a rejection notification.
	// Called when a PM-issued command was rejected.
	OnRejected(
		notification *pb.Notification,
		state S,
		targetDomain string,
		targetCommand string,
	) (*RejectionHandlerResponse, error)
}

// ============================================================================
// Projector Handler Interface
// ============================================================================

// ProjectorDomainHandler is the handler for a single domain's events in a projector.
//
// Projectors consume events and produce external output (read models,
// caches, external systems).
//
// Example:
//
//	type PlayerProjectorHandler struct{}
//
//	func (h *PlayerProjectorHandler) EventTypes() []string {
//	    return []string{"examples.PlayerRegistered", "examples.FundsDeposited"}
//	}
//
//	func (h *PlayerProjectorHandler) Project(events *pb.EventBook) (*pb.Projection, error) {
//	    // Update external read model
//	}
type ProjectorDomainHandler interface {
	// EventTypes returns the fully-qualified event type names this handler processes.
	EventTypes() []string

	// Project processes events and produces external output.
	Project(events *pb.EventBook) (*pb.Projection, error)
}

File: client/rust/src/router/traits.rs
//! Handler traits for each component type.
//!
//! Each trait defines the contract for domain handlers. Implementors
//! encapsulate their routing logic internally and declare which types
//! they handle via `command_types()` or `event_types()`.

use prost_types::Any;
use std::error::Error;
use tonic::Status;

use crate::proto::{CommandBook, Cover, EventBook, Notification, Projection};
use crate::router::StateRouter;

// ============================================================================
// Common Types
// ============================================================================

/// Error type for command/event rejection with a human-readable reason.
#[derive(Debug, Clone)]
pub struct CommandRejectedError {
    pub reason: String,
}

impl CommandRejectedError {
    pub fn new(reason: impl Into<String>) -> Self {
        Self {
            reason: reason.into(),
        }
    }
}

impl std::fmt::Display for CommandRejectedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Command rejected: {}", self.reason)
    }
}

impl std::error::Error for CommandRejectedError {}

impl From<CommandRejectedError> for Status {
    fn from(err: CommandRejectedError) -> Self {
        Status::failed_precondition(err.reason)
    }
}

/// Result type for handlers.
pub type CommandResult<T> = std::result::Result<T, CommandRejectedError>;

/// Response from rejection handlers.
///
/// Handlers may return:
/// - Events to compensate/fix state
/// - Notification to forward upstream
/// - Both
#[derive(Default)]
pub struct RejectionHandlerResponse {
    /// Events to persist (compensation).
    pub events: Option<EventBook>,
    /// Notification to forward upstream.
    pub notification: Option<Notification>,
}

/// Response from saga handlers.
#[derive(Default)]
pub struct SagaHandlerResponse {
    /// Commands to send to other aggregates.
    pub commands: Vec<CommandBook>,
    /// Facts/events to inject to other aggregates.
    pub events: Vec<EventBook>,
}

/// Response from process manager handlers.
#[derive(Default)]
pub struct ProcessManagerResponse {
    /// Commands to send to other aggregates.
    pub commands: Vec<CommandBook>,
    /// Events to persist to the PM's own domain.
    pub process_events: Option<EventBook>,
    /// Facts to inject to other aggregates.
    pub facts: Vec<EventBook>,
}

/// Helper trait for unpacking Any messages.
pub trait UnpackAny {
    /// Unpack an Any to a specific message type.
    fn unpack<M: prost::Message + Default>(&self) -> Result<M, prost::DecodeError>;
}

impl UnpackAny for Any {
    fn unpack<M: prost::Message + Default>(&self) -> Result<M, prost::DecodeError> {
        M::decode(self.value.as_slice())
    }
}

// ============================================================================
// Command Handler
// ============================================================================

/// Handler for a single domain's command handler logic.
///
/// Command handlers receive commands and emit events. They maintain state
/// that is rebuilt from events using a `StateRouter`.
///
/// # Example
///
/// ```rust,ignore
/// struct PlayerHandler {
///     state_router: StateRouter<PlayerState>,
/// }
///
/// impl CommandHandlerDomainHandler for PlayerHandler {
///     type State = PlayerState;
///
///     fn command_types(&self) -> Vec<String> {
///         vec!["RegisterPlayer".into(), "DepositFunds".into()]
///     }
///
///     fn state_router(&self) -> &StateRouter<Self::State> {
///         &self.state_router
///     }
///
///     fn handle(
///         &self,
///         cmd: &CommandBook,
///         payload: &Any,
///         state: &Self::State,
///         seq: u32,
///     ) -> CommandResult<EventBook> {
///         dispatch_command!(payload, cmd, state, seq, {
///             "RegisterPlayer" => self.handle_register,
///             "DepositFunds" => self.handle_deposit,
///         })
///     }
/// }
/// ```
pub trait CommandHandlerDomainHandler: Send + Sync {
    /// The state type for this aggregate.
    type State: Default + 'static;

    /// Command type suffixes this handler processes.
    ///
    /// Used for subscription derivation and routing.
    fn command_types(&self) -> Vec<String>;

    /// Get the state router for rebuilding state from events.
    fn state_router(&self) -> &StateRouter<Self::State>;

    /// Rebuild state from events.
    ///
    /// Default implementation uses `state_router().with_event_book()`.
    fn rebuild(&self, events: &EventBook) -> Self::State {
        self.state_router().with_event_book(events)
    }

    /// Handle a command and return resulting events.
    ///
    /// The handler should dispatch internally based on `payload.type_url`.
    fn handle(
        &self,
        cmd: &CommandBook,
        payload: &Any,
        state: &Self::State,
        seq: u32,
    ) -> CommandResult<EventBook>;

    /// Handle a rejection notification.
    ///
    /// Called when a command issued by a saga/PM targeting this aggregate's
    /// domain was rejected. Override to provide custom compensation logic.
    ///
    /// Default implementation returns an empty response (framework handles).
    fn on_rejected(
        &self,
        _notification: &Notification,
        _state: &Self::State,
        _target_domain: &str,
        _target_command: &str,
    ) -> CommandResult<RejectionHandlerResponse> {
        Ok(RejectionHandlerResponse::default())
    }
}

// ============================================================================
// Saga Handler
// ============================================================================

/// Handler for a single domain's events in a saga.
///
/// Sagas are **pure translators**: they receive source events and produce
/// commands for target domains. They do NOT receive destination state —
/// the framework handles sequence stamping and delivery retries.
///
/// # Example
///
/// ```rust,ignore
/// struct OrderSagaHandler;
///
/// impl SagaDomainHandler for OrderSagaHandler {
///     fn event_types(&self) -> Vec<String> {
///         vec!["OrderCompleted".into(), "OrderCancelled".into()]
///     }
///
///     fn handle(
///         &self,
///         source: &EventBook,
///         event: &Any,
///     ) -> CommandResult<SagaHandlerResponse> {
///         dispatch_event!(event, source, {
///             "OrderCompleted" => self.handle_completed,
///             "OrderCancelled" => self.handle_cancelled,
///         })
///     }
/// }
/// ```
pub trait SagaDomainHandler: Send + Sync {
    /// Event type suffixes this handler processes.
    ///
    /// Used for subscription derivation.
    fn event_types(&self) -> Vec<String>;

    /// Translate source events into commands for target domains.
    ///
    /// Commands should have `cover` set to identify the target aggregate.
    /// The framework will stamp `angzarr_deferred` with source info and
    /// handle sequence assignment on delivery.
    ///
    /// Returns commands to send to other aggregates and events/facts to inject.
    fn handle(&self, source: &EventBook, event: &Any) -> CommandResult<SagaHandlerResponse>;

    /// Handle a rejection notification.
    ///
    /// Called when a saga-issued command was rejected. Override to provide
    /// custom compensation logic.
    ///
    /// Default implementation returns an empty response (framework handles).
    fn on_rejected(
        &self,
        _notification: &Notification,
        _target_domain: &str,
        _target_command: &str,
    ) -> CommandResult<RejectionHandlerResponse> {
        Ok(RejectionHandlerResponse::default())
    }
}

// ============================================================================
// Process Manager Handler
// ============================================================================

/// Handler for a single domain's events in a process manager.
///
/// Process managers correlate events across multiple domains and maintain
/// their own state. Each domain gets its own handler, but they all share
/// the same PM state type.
///
/// # Example
///
/// ```rust,ignore
/// struct OrderPmHandler;
///
/// impl ProcessManagerDomainHandler<HandFlowState> for OrderPmHandler {
///     fn event_types(&self) -> Vec<String> {
///         vec!["OrderCreated".into()]
///     }
///
///     fn prepare(&self, trigger: &EventBook, state: &HandFlowState, event: &Any) -> Vec<Cover> {
///         // Declare needed destinations
///         vec![]
///     }
///
///     fn handle(
///         &self,
///         trigger: &EventBook,
///         state: &HandFlowState,
///         event: &Any,
///         destinations: &[EventBook],
///     ) -> CommandResult<ProcessManagerResponse> {
///         // Process event, emit commands and/or PM events
///         Ok(ProcessManagerResponse::default())
///     }
/// }
/// ```
pub trait ProcessManagerDomainHandler<S>: Send + Sync {
    /// Event type suffixes this handler processes.
    fn event_types(&self) -> Vec<String>;

    /// Prepare phase — declare destination covers needed.
    fn prepare(&self, trigger: &EventBook, state: &S, event: &Any) -> Vec<Cover>;

    /// Handle phase — produce commands and PM events.
    fn handle(
        &self,
        trigger: &EventBook,
        state: &S,
        event: &Any,
        destinations: &[EventBook],
    ) -> CommandResult<ProcessManagerResponse>;

    /// Handle a rejection notification.
    ///
    /// Called when a PM-issued command was rejected. Override to provide
    /// custom compensation logic.
    fn on_rejected(
        &self,
        _notification: &Notification,
        _state: &S,
        _target_domain: &str,
        _target_command: &str,
    ) -> CommandResult<RejectionHandlerResponse> {
        Ok(RejectionHandlerResponse::default())
    }
}

// ============================================================================
// Projector Handler
// ============================================================================

/// Handler for a single domain's events in a projector.
///
/// Projectors consume events and produce external output (read models,
/// caches, external systems).
///
/// # Example
///
/// ```rust,ignore
/// struct PlayerProjectorHandler;
///
/// impl ProjectorDomainHandler for PlayerProjectorHandler {
///     fn event_types(&self) -> Vec<String> {
///         vec!["PlayerRegistered".into(), "FundsDeposited".into()]
///     }
///
///     fn project(&self, events: &EventBook) -> Result<Projection, Box<dyn Error + Send + Sync>> {
///         // Update external read model
///         Ok(Projection::default())
///     }
/// }
/// ```
pub trait ProjectorDomainHandler: Send + Sync {
    /// Event type suffixes this handler processes.
    fn event_types(&self) -> Vec<String>;

    /// Project events to external output.
    fn project(&self, events: &EventBook) -> Result<Projection, Box<dyn Error + Send + Sync>>;
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_rejected_error_display() {
        let err = CommandRejectedError::new("insufficient funds");
        assert_eq!(err.to_string(), "Command rejected: insufficient funds");
    }

    #[test]
    fn command_rejected_error_to_status() {
        let err = CommandRejectedError::new("invalid input");
        let status: Status = err.into();
        assert_eq!(status.code(), tonic::Code::FailedPrecondition);
    }

    #[test]
    fn rejection_handler_response_default() {
        let response = RejectionHandlerResponse::default();
        assert!(response.events.is_none());
        assert!(response.notification.is_none());
    }

    #[test]
    fn process_manager_response_default() {
        let response = ProcessManagerResponse::default();
        assert!(response.commands.is_empty());
        assert!(response.process_events.is_none());
        assert!(response.facts.is_empty());
    }

    #[test]
    fn saga_handler_response_default() {
        let response = SagaHandlerResponse::default();
        assert!(response.commands.is_empty());
        assert!(response.events.is_empty());
    }
}
